pkg/controller/podgc/queue.go

```diff
@@ -0,0 +1,69 @@
+/*
+Copyright 2015 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package podgc
+
+import (
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/clock"
+)
+
+// UniqueWorkQueue allows queuing items with a timestamp. An item is
+// considered ready to process if the timestamp has expired.
+type UniqueWorkQueue interface {
```

Comment:
I'm generally not a fan of creating new structures. 
I think we can simply use DelayingQueue:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go#L37

Basically it gives you "Get()" and "AddAfter()" methods, which seem to be all you need.

Reply:
DelayingQueue does not have ability to deduplicate items.

Reply:
IIRC - it's impossible to have mutliple items in Queue (and thus DelayingQueue) with the same key at the same time.
So it's kind-of deduping out-of-the-box.

Reply:
+1, could you rewrite to DelayingQueue?
As Wojtek said, it dedupes items, there is even a test for that - https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/util/workqueue/delaying_queue_test.go#L72

Reply:
Done